core/system/src/register_gateway.c

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright (c) 2016, ARM Limited, All Rights Reserved
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <uvisor.h>
+#include "halt.h"
+#include "register_gateway.h"
+#include "vmpu.h"
+
+/** Perform the validation of a register gateway.
+ * @internal
+ * @warning This function does not verify that the operation, value and mask
+ * make sense together. The code that performs the operation should check for
+ * this and return an error accordingly.
+ * @param register_gateway[in]  Pointer to the register gateway structure in
+ *                              flash.
+ * @returns the extracted address of the register gateway.
+ */
+static TRegisterGateway * register_gateway_check(TRegisterGateway * register_gateway)
+{
+    /* Verify that the register gateway is in public flash. */
+    /* Note: We assume here that TRegisterGateway is a multiple of 32 bit. If
+     * that is not true there might be false negatives with an unaligned secure
+     * gateway. */
+    int word = 0;
+    for (; word < sizeof(TRegisterGateway) / sizeof(uint32_t); word++) {
+        if (!vmpu_public_flash_addr(((uint32_t) register_gateway) + sizeof(uint32_t) * word)) {
+            HALT_ERROR(SANITY_CHECK_FAILED, "Register gateway 0x%08X is not in public flash.",
+                       (uint32_t) register_gateway);
+        }
+    }
```

Comment:
@AlessandroA: too heavy-handed - please just verify the alignment of the start and that the start is a vmpu_public_flash_addr